File: Python/BFS/M-994. Rotting Oranges.py
from collections import deque


class Solution:
    @classmethod
    def orangesRotting(self, grid: list[list[int]]) -> int:
        FRESH, ROTTEN = 1, 2
        m, n = len(grid), len(grid[0])
        rotten_orange_queue = deque([])
        fresh_oranges = 0

        for i in range(m):
            for j in range(n):
                if grid[i][j] == ROTTEN:
                    rotten_orange_queue.append([i, j, 0])
                elif grid[i][j] == FRESH:
                    fresh_oranges += 1
        if fresh_oranges == 0:
            return 0

        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

        def is_within_border(row: int, col: int):
            return row < m and row >= 0 and col < n and col >= 0

        while rotten_orange_queue:
            row, col, minutes = rotten_orange_queue.popleft()
            for row_dist, col_dist in directions:
                moved_row, moved_col = row + row_dist, col + col_dist
                if (
                    is_within_border(moved_row, moved_col)
                    and grid[moved_row][moved_col] == FRESH
                ):
                    grid[moved_row][moved_col] = ROTTEN
                    fresh_oranges -= 1
                    if fresh_oranges == 0:
                        return minutes + 1
                    rotten_orange_queue.append([moved_row, moved_col, minutes + 1])
        return -1


testData = [
    {
        "grid": [
            [2, 1, 1],
            [1, 1, 0],
            [0, 1, 1],
        ],
        "ans": 4,
    },
    {
        "grid": [
            [2, 1, 1],
            [0, 1, 1],
            [1, 0, 1],
        ],
        "ans": -1,
    },
    {
        "grid": [[0, 2]],
        "ans": 0,
    },
]

for data in testData:
    print(data["ans"] == Solution.orangesRotting(data["grid"]))

File: BFS/M-994. Rotting Oranges/M-994. Rotting Oranges.ts
function orangesRotting(grid: number[][]): number {
  const [m, n] = [grid.length, grid[0].length];
  let freshOranges: number = 0;
  const rottenOrangeQueue: number[][] = [];

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] === 2) rottenOrangeQueue.push([i, j, 0]);
      else if (grid[i][j] === 1) freshOranges++;
    }
  }

  if (freshOranges === 0) return 0;

  const directions = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ];
  const isWithinBorder = (row: number, col: number): boolean => {
    return row >= 0 && row < m && col >= 0 && col < n;
  };
  while (rottenOrangeQueue.length) {
    const [row, col, minute] = rottenOrangeQueue.shift() as number[];
    for (const [rowDist, colDist] of directions) {
      const [movedRow, movedCol] = [row + rowDist, col + colDist];
      if (
        isWithinBorder(movedRow, movedCol) &&
        grid[movedRow][movedCol] === 1
      ) {
        grid[movedRow][movedCol] = 2;
        if (--freshOranges == 0) return minute + 1;
        rottenOrangeQueue.push([movedRow, movedCol, minute + 1]);
      }
    }
  }

  return -1;
}

const testData = [
  {
    grid: [
      [2, 1, 1],
      [1, 1, 0],
      [0, 1, 1],
    ],
    ans: 4,
  },
  {
    grid: [
      [2, 1, 1],
      [0, 1, 1],
      [1, 0, 1],
    ],
    ans: -1,
  },
  {
    grid: [[0, 2]],
    ans: 0,
  },
];

for (const data of testData) {
  console.log(data['ans'] === orangesRotting(data['grid']));
}
